### PR TITLE
add some func for validation

### PR DESCRIPTION
**Pull request by wolfstudy**
add some func for validation

### CODE REVIEW

**Comment by wolfstudy** [comment]
@jjz @ludete PTAL

File: blockchain/Validation.go [hunk 1]
```diff
@@ -2468,6 +2468,92 @@ func ContextualCheckTransactionForCurrentBlock(tx *model.Tx, state *model.Valida
...
+		return false
+	}
+	return true
+}
+
+// UnloadBlockIndex may not be used after any connections are up as much of the peer-processing
+// logic assumes a consistent block index state
+func UnloadBlockIndex() {
+	//TODO:LOCK(cs_main);
+	var c *ChainState
```
Comment (ludete) [inline blockchain/Validation.go]: this should be global variable

File: blockchain/Validation.go [hunk 2]
```diff
@@ -2468,6 +2468,92 @@ func ContextualCheckTransactionForCurrentBlock(tx *model.Tx, state *model.Valida
...
+	c.setBlockIndexCandidates.End()
+	GChainActive.SetTip(nil)
+	gpindexBestInvalid = nil
+	gpindexBestHeader = nil
+	gmpool.Clear()
+	c.MapBlocksUnlinked = nil
+	ginfoBlockFile = nil
```
Comment (ludete) [inline blockchain/Validation.go]: should not use ginfoBlockFile = nil

File: blockchain/Validation.go [hunk 3]
```diff
@@ -2468,6 +2468,92 @@ func ContextualCheckTransactionForCurrentBlock(tx *model.Tx, state *model.Valida
...
+	gLastBlockFile = 0
+	gnBlockSequenceID = 1
+	gsetDirtyFileInfo.Clear()
+	gsetDirtyBlockIndex.Clear()
+	versionBitsCache.Clear()
+	for b := 0; b < VERSIONBITS_NUM_BITS; b++ {
+		warningcache[b] = nil
```
Comment (ludete) [inline blockchain/Validation.go]: should not use nil to initialization map

File: blockchain/Validation.go [hunk 4]
```diff
@@ -2468,6 +2468,92 @@ func ContextualCheckTransactionForCurrentBlock(tx *model.Tx, state *model.Valida
...
+	}
+
+	for entry := range MapBlockIndex.Data {
+		delete(MapBlockIndex.Data, entry)
+	}
+	MapBlockIndex.Data = nil
```
Comment (ludete) [inline blockchain/Validation.go]: the map struct initialization should not use nil

File: blockchain/Validation.go [hunk 5]
```diff
@@ -2468,6 +2468,92 @@ func ContextualCheckTransactionForCurrentBlock(tx *model.Tx, state *model.Valida
...
+	gLastBlockFile = 0
+	gnBlockSequenceID = 1
+	gsetDirtyFileInfo.Clear()
+	gsetDirtyBlockIndex.Clear()
+	versionBitsCache.Clear()
+	for b := 0; b < VERSIONBITS_NUM_BITS; b++ {
+		warningcache[b] = nil
+	}
+
+	for entry := range MapBlockIndex.Data {
```
Comment (ludete) [inline blockchain/Validation.go]: here shoule ingore, if you use Go languange.

Comment (wolfstudy) [reply blockchain/Validation.go]: OK， get it

File: blockchain/Validation.go (hunk 2 above)
Comment (wolfstudy) [reply blockchain/Validation.go]: the slice zero is nil or remake slice,

File: blockchain/Validation.go (hunk 3 above)
Comment (wolfstudy) [reply blockchain/Validation.go]: in order to warningcache
（`var warningcache [VERSIONBITS_NUM_BITS]ThresholdConditionCache`）is nil, i think the method is ok

File: blockchain/Validation.go (hunk 4 above)
Comment (wolfstudy) [reply blockchain/Validation.go]: the map zero is nil, i think this ok

**Comment by wolfstudy** [comment]
@ludete PTAL again , thanks

**Comment by wolfstudy** [comment]
@ludete @jjz PTAL

File: blockchain/Validation.go [hunk 6]
```diff
@@ -2920,9 +3320,92 @@ func CheckInputs(tx *model.Tx, state *model.ValidationState, cache *utxo.CoinsVi
...
+			// CheckSequenceLocks to indicate the LockPoints validity
+			maxInputHeight := 0
+			for height := range prevheights {
+				// Can ignore mempool inputs since we'll fail if they had non-zero locks
+				if height != tip.Height+1 {
+					maxInputHeight = int(math.Max(float64(maxInputHeight), float64(height)))
+				}
+			}
+			//todo:mempool not define maxInputBlock: lp.maxInputBlock=tip.GetAncestor(maxInputHeight)
+			tip.GetAncestor(maxInputHeight)
```
Comment (ludete) [inline blockchain/Validation.go]: loss code

Comment (wolfstudy) [reply blockchain/Validation.go]: ok, get it

File: blockchain/Validation.go [hunk 7]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+
 func ConnectBlock(param *msg.BitcoinParams, pblock *model.Block, state *model.ValidationState,
 	pindex *BlockIndex, view *utxo.CoinsViewCache, fJustCheck bool) bool {
+
+	//TODO: AssertLockHeld(cs_main);
+	//var sc sync.RWMutex
+	//sc.Lock()
+	//defer sc.Unlock()
+
+	nTimeStart := utils.GetMockTimeInMicros()
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: yes, i should call the `utils.GetMicrosTime()` not `utils.GetMockTimeInMicros()`

File: blockchain/Validation.go [hunk 8]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+
+	// Check it again in case a previous version let a bad block in
+	if !CheckBlock(param, pblock, state, !fJustCheck, !fJustCheck) {
+		log.Error("CheckBlock: %s", FormatStateMessage(state))
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: ok, i will use `logger.ErrorLog(fmt.Sprintf("CheckBlock: %s", FormatStateMessage(state)))`

File: blockchain/Validation.go [hunk 9]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+	}
+
+	// Verify that the view's current state corresponds to the previous block
+	var hashPrevBlock utils.Hash
+	if pindex.PPrev == nil {
+		hashPrevBlock = utils.Hash{}
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: yes, if and else return same value.

File: blockchain/Validation.go [hunk 10]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+		hashPrevBlock = pindex.PPrev.GetBlockHash()
+	}
+
+	if hashPrevBlock != view.GetBestBlock() {
+		panic("error: hashPrevBlock not equal view.GetBestBlock()")
+	}
+
+	// Special case for the genesis block, skipping connection of its
+	// transactions (its coinbase is unspendable)
+	if &pblock.Hash == param.GenesisHash {
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: i should use `pblock.Hash.Cmp(param.GenesisHash) != 0` instead `&pblock.Hash == param.GenesisHash`

[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: i should `pblock.Hash.IsEqual(param.GenesisHash)` instead `&pblock.Hash == param.GenesisHash`

File: blockchain/Validation.go [hunk 11]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+		if !fJustCheck {
+			view.SetBestBlock(pindex.GetBlockHash())
+		}
+		return true
+	}
+
+	fScriptChecks := true
+	if &HashAssumeValid != nil {
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: yes ,the HashAssumeValid address always not nil

File: blockchain/Validation.go [hunk 12]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+			// further back. The test against nMinimumChainWork prevents the
+			// skipping when denied access to any chain at least as good as
+			// the expected chain.
+			fScriptChecks = (GetBlockProofEquivalentTime(gpindexBestHeader, pindex, gpindexBestHeader, param)) <= 60*60*24*7*2
+		}
+	}
+
+	nTime1 := utils.GetMockTimeInMicros()
+	gnTimeCheck += nTime1 - nTimeStart
+	log.Info("bench", "    - Sanity checks: %.2fms [%.2fs]\n", 0.001*float64(nTime1-nTimeStart), float64(gnTimeCheck)*0.000001)
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: i should use `logger.GetLogger().Info()` instead `log.Info()`

File: blockchain/Validation.go [hunk 13]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+	// Once BIP34 activated it was not possible to create new duplicate
+	// coinbases and thus other than starting with the 2 existing duplicate
+	// coinbase pairs, not possible to create overwriting txs. But by the time
+	// BIP34 activated, in each of the existing pairs the duplicate coinbase had
+	// overwritten the first before the first had been spent. Since those
+	// coinbases are sufficiently buried its no longer possible to create
+	// further duplicate transactions descending from the known pairs either. If
+	// we're on the known chain at height greater than where BIP34 activated, we
+	// can save the db accesses needed for the BIP30 check.
+	var pindexBIP34height *BlockIndex
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: already fix in under commit

File: blockchain/Validation.go [hunk 14]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+	}
+
+	// Start enforcing BIP68 (sequence locks) using versionbits logic.
+	nLockTimeFlags := 0
+	if VersionBitsState(pindex.PPrev, param, msg.DEPLOYMENT_CSV, &versionBitsCache) == THRESHOLD_ACTIVE {
+		nLockTimeFlags |= consensus.LocktimeVerifySequence
+	}
+
+	flags := GetBlockScriptFlags(pindex, param)
+	nTime2 := utils.GetMockTimeInMicros()
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: get it

File: blockchain/Validation.go [hunk 15]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+			// doing any signature verification at all in testing without
+			// having to artificially set the default assumed verified block
+			// further back. The test against nMinimumChainWork prevents the
+			// skipping when denied access to any chain at least as good as
+			// the expected chain.
+			fScriptChecks = (GetBlockProofEquivalentTime(gpindexBestHeader, pindex, gpindexBestHeader, param)) <= 60*60*24*7*2
+		}
+	}
+
+	nTime1 := utils.GetMockTimeInMicros()
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: get it

File: blockchain/Validation.go [hunk 16]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+	fScriptChecks := true
+	if &HashAssumeValid != nil {
+		// We've been configured with the hash of a block which has been
+		// externally verified to have a valid history. A suitable default value
+		// is included with the software and updated from time to time. Because
+		// validity relative to a piece of software is an objective fact these
+		// defaults can be easily reviewed. This setting doesn't force the
+		// selection of any particular chain but makes validating some faster by
+		// effectively caching the result of part of the verification.
+		it := MapBlockIndex.Data[HashAssumeValid]
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: get it

File: blockchain/Validation.go [hunk 17]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+	// overwritten the first before the first had been spent. Since those
+	// coinbases are sufficiently buried its no longer possible to create
+	// further duplicate transactions descending from the known pairs either. If
+	// we're on the known chain at height greater than where BIP34 activated, we
+	// can save the db accesses needed for the BIP30 check.
+	var pindexBIP34height *BlockIndex
+	pindexBIP34height = pindex.PPrev.GetAncestor(param.BIP34Height)
+	// Only continue to enforce if we're below BIP34 activation height or the
+	// block hash at that height doesn't correspond.
+	fEnforceBIP30 = fEnforceBIP30 && (&pindexBIP34height == nil || !(pindexBIP34height.GetBlockHash() == param.BIP34Hash))
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: use `pindex.PHashBlock != utils.HashZero` instead `&pindexBIP34height == nil`

File: blockchain/Validation.go [hunk 18]
```diff
@@ -1538,8 +1543,324 @@ func InvalidBlockFound(pindex *BlockIndex, state *model.ValidationState) {
...
+
+	if fJustCheck {
+		return true
+	}
+
+	// Write undo information to disk
+	tmpUndoPos := pindex.GetUndoPos()
+	if tmpUndoPos.IsNull() || !pindex.IsValid(BLOCK_VALID_SCRIPTS) {
+		if tmpUndoPos.IsNull() {
+			var pos *DiskBlockPos
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: get it

File: blockchain/Validation.go [hunk 19]
```diff
@@ -1652,6 +1973,34 @@ func DisconnectBlock(pblock *model.Block, pindex *BlockIndex, view *utxo.CoinsVi
...
 
+func UndoWriteToDisk(blockundo *BlockUndo, pos *DiskBlockPos, hashBlock utils.Hash, messageStart btcutil.BitcoinNet) bool {
+	// Open history file to append
+	fileout := OpenUndoFile(*pos, false)
+	if fileout == nil {
+		return logger.ErrorLog("OpenUndoFile failed")
+	}
+
+	// Write index header
+	nSize := 0
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: ok, the func is not finish

File: blockchain/Validation.go [hunk 20]
```diff
@@ -1652,6 +1973,34 @@ func DisconnectBlock(pblock *model.Block, pindex *BlockIndex, view *utxo.CoinsVi
...
+	err := utils.BinarySerializer.PutUint32(fileout, binary.LittleEndian, uint32(messageStart))
+	if err != nil {
+		logger.ErrorLog("the messageStart write failed")
+	}
+	utils.WriteVarInt(fileout, uint64(nSize))
+
+	// Write undo data
+	fileOutPos, err := fileout.Seek(0, 1)
+	if fileOutPos < 0 || err != nil {
+		logger.ErrorLog("UndoWriteToDisk: ftell failed")
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: get it

File: blockchain/Validation.go [hunk 21]
```diff
@@ -2910,6 +3247,69 @@ func CheckInputs(tx *model.Tx, state *model.ValidationState, cache *utxo.CoinsVi
...
+	// time constraints given our view of block chain history.
+	// The semantics of nLockTime are the last invalid height/time, so
+	// use -1 to have the effect of any height or time being valid.
+
+	nMinHeight := -1
+	nMinTime := -1
+	// tx.nVersion is signed integer so requires cast to unsigned otherwise
+	// we would be doing a signed comparison and half the range of nVersion
+	// wouldn't support BIP 68.
+	fEnforceBIP68 := tx.Version >= 2 && flags != 0&consensus.LocktimeVerifySequence
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: get it

File: blockchain/Validation.go [hunk 22]
```diff
@@ -2910,6 +3247,69 @@ func CheckInputs(tx *model.Tx, state *model.ValidationState, cache *utxo.CoinsVi
...
+		maps[nMinHeight] = int64(nMinTime)
+		return maps
+	}
+
+	for txinIndex := 0; txinIndex < len(tx.Ins); txinIndex++ {
+		txin := tx.Ins[txinIndex]
+		// Sequence numbers with the most significant bit set are not
+		// treated as relative lock-times, nor are they given any
+		// consensus-enforced meaning at this point.
+		if txin.Sequence != 0&model.SEQUENCE_LOCKTIME_DISABLE_FLAG {
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: get it

File: blockchain/Validation.go [hunk 23]
```diff
@@ -2910,6 +3247,69 @@ func CheckInputs(tx *model.Tx, state *model.ValidationState, cache *utxo.CoinsVi
...
+			// The height of this input is not relevant for sequence locks
+			prevHeights[txinIndex] = 0
+			continue
+		}
+		nCoinHeight := prevHeights[txinIndex]
+
+		if txin.Sequence != 0&model.SEQUENCE_LOCKTIME_DISABLE_FLAG {
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply blockchain/Validation.go]: get it

File: blockchain/Validation.go [hunk 24]
```diff
@@ -1594,66 +1594,63 @@ func ConnectBlock(param *msg.BitcoinParams, pblock *model.Block, state *model.Va
...
-			// having to artificially set the default assumed verified block
-			// further back. The test against nMinimumChainWork prevents the
-			// skipping when denied access to any chain at least as good as
-			// the expected chain.
-			fScriptChecks = (GetBlockProofEquivalentTime(gpindexBestHeader, pindex, gpindexBestHeader, param)) <= 60*60*24*7*2
-		}
-	}
-
-	nTime1 := utils.GetMockTimeInMicros()
+		if MapBlockIndex.Data[HashAssumeValid] != nil {
```
Comment (ludete) [inline blockchain/Validation.go]: why do you not use if v , ok := MapBlockIndex.Data[HashAssumeValid]; ok{
.....
}

Comment (wolfstudy) [reply blockchain/Validation.go]: fix alerady

**Comment by ludete** [comment]
OK

**Pull request merged**